### ISSUE TITLE
Reverse proxy sends no body on retry after 503 service unavailable

### ISSUE THREAD

**Issue by ghelyar** [Azure/service-fabric-issues#381]
I have an asp.net core REST service with a POST action which returns 503 Service Unavailable with a Retry-After header when it is busy.

When Service Fabric Reverse Proxy retries this, the controller doesn't get the request until it's timed out. After the timeout, the controller gets an empty [FromBody] and the CancellationToken has been set. When reading it as a middleware, context.Request.Body.ReadByte() works on the initial request but blocks on the retry until it times out with econnreset.

This happens on a local cluster with reverse proxy enabled in the cluster manifest at version 5.6.220.9494

Repro:
Server:
```
app.Run(async context => {
  var b = context.Request.Body.ReadByte();
  context.Response.StatusCode = (int)HttpStatusCode.ServiceUnavailable;
  context.Response.Headers.Add("Retry-After", "3");
});
```

Client:
```
var client = new HttpClient();
var response = await client.PostAsync("http://localhost:19081/AppName/ServiceName/", new StringContent("test", Encoding.UTF8, "text/plain"));
response.EnsureSuccessStatusCode();
var content = await response.Content.ReadAsStringAsync();
```

On the first request, the server returns 503 with the Retry-After header.
On the second request, the server blocks on ReadByte and the client eventually throws HttpRequestException 504 (Reverse Proxy Timeout)

**Comment by kavyako** [comment]
Yes this is a bug in the reverse proxy.
While forwarding the request for the first time, reverse proxy reads the body from the client's request and forwards it. We do not cache it internally. Upon retrying with non idempotent verbs like POST, this body is lost and cannot be resent.
We will fix the retry policy in the upcoming 5.7 CU, to not retry for non idempotent verbs and let the clients handle the retry logic.
For future releases we could look at doing something more intelligent like retry on certain conditions even for POST requests, that requires more thought.

**Comment by ghelyar** [comment]
While I'm fine with it not retrying, and ideally it would be nice to be able to customise the retry behaviour so that it can do things like never transparently retry, I think some care needs to be taken here.

While GET _should be_ idempotent, not all gets are, and POSTs _can be_ idempotent. For example, I have a stateless service that does some deterministic processing on some data with no side effects and returns the processed result. I need to use a verb that allows me to submit the input data.

I would also say that if the server has returned a retry-after header, that probably means that it's safe to retry.

There is also the 401 Gone response when a service has moved between nodes. I'm not sure if the reverse proxy needs to handle that, but if it does then it should be able to retry safely.

If a retry does happen, then the server should not block waiting for a body that will never come. I don't know if this is a problem with asp.net core or a malformed request from the reverse proxy e.g. there was certainly a content length with no body.

**Comment by kavyako** [comment]
This has been fixed in 6.0, reverse proxy will not retry on POST for the scenario mentioned here, so you should not see a blocking retry.

Here is the current retry behavior:

For service response status codes 404 (without X-ServiceFabric=ResourceNotFound), 410 and 503, reverse proxy will retry the request only for idempotent safe methods: GET, OPTIONS and HEAD. For any other HTTP method, it will not retry and forward the status code back to the client.
As of now, Reverse proxy will not honor the Retry-After header for such retries. 

I am closing this issue for now since we will not be adding any new retry functionality/feature support in the immediate releases.
